[Crypto] optimize sha3 in the crypto library

tarakby opened the pull request:
Addresses https://github.com/dapperlabs/flow-go/issues/5342

The Go standard crypto lib implements a generic keccak algorithm that is common for SHA3 (all output versions) and SHAKE. It also provides a flexible API that allows a few features that don't seem to be used by Flow (continue writing in a hasher after a hash was computed for instance).
This PR customizes the Go standard implementation to only support SHA3-256 and SHA3-384, and removes the API flexibility to only keep basic functionalities.

This results in copying a lot of the Go standard library, since most of the functions are private. This also means Flow lib will need to maintain its own implementation of sha3 and keep an eye on future SHA3 implementation updates. 

The changes reduce the computation time, the memory usage, and number of allocations:

Before
```
BenchmarkSha3_256-8             767 ns/op             512 B/op          3 allocs/op
BenchmarkSha3_384-8             746 ns/op             544 B/op          3 allocs/op
```

after
```
BenchmarkComputeHash/SHA3_256-8                 510 ns/op              32 B/op          1 allocs/op
BenchmarkComputeHash/SHA3_384-8                 498 ns/op              48 B/op          1 allocs/op
```

ramtinms reviewed (approved):
Looks good to me,

AlexHentschel commented on crypto/hash/sha3.go:
```diff
@@ -1,65 +1,193 @@
+// calling Reset().
+func (s *sha3State) ComputeHash(data []byte) Hash {
 	s.Reset()
 	_, _ = s.Write(data)
```
I would be almost tempted to change https://github.com/onflow/flow-go/blob/dd7a60344577faf159bfec0ec55c32bcdc4bf62e/crypto/hash/sha3.go#L131
to 
```golang
func (d *sha3State) Write(p []byte) (written int, err error) {
    return d.write(p), nil
}

func (d *sha3State) write(p []byte) written int {
```
and call `s.write(data)` here ...

AlexHentschel commented on engine/execution/state/delta/view.go:
```diff
@@ -253,8 +250,15 @@ func (v *View) ReadsCount() uint64 {
+// Once called, it doesn't allow writing more data into the SPoCK secret.
 func (v *View) SpockSecret() []byte {
-	return v.spockSecretHasher.SumHash()
+	// check if spocksSecret has been already computed
+	if v.spocksSecret == nil {
+		v.spocksSecret = v.spockSecretHasher.SumHash()
+	}
+	return v.spocksSecret
```
Are we sure that this does not break any concurrency?

tarakby replied:
Not sure if we the function needs to be thread safe, but I added a lock in case it needs to be.

AlexHentschel reviewed (approved):
👍

tarakby commented on crypto/hash/hash.go:
```diff
@@ -26,44 +26,19 @@ func (h Hash) String() string {
 
 // Hasher interface
 type Hasher interface {
```
Good point!

tarakby commented:
Slowing down on addressing the comments and merging this PR as I would like to get https://github.com/dapperlabs/flow-go/issues/5563 done first. https://github.com/dapperlabs/flow-go/issues/5563 would allow testing non-amd64 code that is introduced by this PR.

huitseeker commented:
> The Go standard crypto lib implements a generic keccak algorithm [...] It also provides a flexible API that allows a few features that don't seem to be used by Flow (_continue writing in a hasher after a hash was computed for instance_).

Am I correct in understanding that what we're not using is
-  the ability to manifest a hash at a partial point in a sequence of updates to a hasher?
- rather than the ability to merge in more inptus to a sha3 hasher?

tarakby commented on crypto/hash/hash.go:
```diff
@@ -26,44 +26,19 @@ func (h Hash) String() string {
+	// It may update the state or not depending on the implementation.
 	SumHash() Hash
-	// Reset resets the hash state
+	// Reset resets the hash state.
```
Here I'm not referring to the generic "state" of the hasher software object, I am referring to the inner state of the crypto hashing function, usually referred to as hash state.

tarakby commented on crypto/hash/sha3.go:
```diff
@@ -1,65 +1,193 @@
+	return s.sum()
+}
+
+// The functions below were copied and modified from golang.org/x/crypto/sha3.
```
That code is still very much related to the public functions so I prefer to keep the code in the same file (190 lines is also relatively short). The core functionality of sha3 is called Keccak, and that code is separated into `keccak.go`, `keccakf.go` and `keccak.s`.

simonhf commented:
> This results in copying a lot of the Go standard library, since most of the functions are private.

@tarakby BTW I'm not advocating to implement things like this, and there are definitely pros and cons, but just in case you didn't know, there is a Golang way to access private functions of another package [1].

[1] https://stackoverflow.com/questions/36009327/how-to-access-private-functions-of-another-package-in-golang/36011850
